### PR TITLE
drivers: sensor: update bme280 to new async api

### PR DESCRIPTION
**Pull request by laurenmurphyx64**
Updates BME280 driver to new async API and convert sample.

Depends on #71093 (merged) and #71160 (still waiting) (commit from #71160 is included in this PR, so this is DNM until that is merged).

Depends on RFC https://github.com/zephyrproject-rtos/zephyr/issues/73031

### CODE REVIEW

**Comment by laurenmurphyx64** [comment]
I realized it would make more sense for me to update the BME280 sample at the same time, give me a sec to do that.

File: samples/sensor/bme280/src/main.c [hunk 1]
```diff
@@ -9,15 +9,27 @@
...
  */
-static const struct device *get_bme280_device(void)
-{
-	const struct device *const dev = DEVICE_DT_GET_ANY(bosch_bme280);
+const struct device *const dev = DEVICE_DT_GET_ANY(bosch_bme280);
+
+SENSOR_DT_READ_IODEV(iodev, DT_NODELABEL(bme280),
+		{SENSOR_CHAN_AMBIENT_TEMP, 0},
+		{SENSOR_CHAN_HUMIDITY, 0},
+		{SENSOR_CHAN_PRESS, 0});
```
Comment (laurenmurphyx64) [inline samples/sensor/bme280/src/main.c:27]: @teburd Because there's no guarantee DEVICE_DT_GET_ANY will return the same of multiple BME280 devices as DT_NODELABEL, I'm trying to figure out how to get the label / alias from the existing device struct so I can pass it into SENSOR_DT_READ_IODEV. Is there something simple I'm missing?

File: samples/sensor/bme280/src/main.c [hunk 2]
```diff
@@ -37,23 +49,65 @@ static const struct device *get_bme280_device(void)
...
+		struct sensor_q31_data *hum_data =
+				(struct sensor_q31_data *) hum_buf;
 
-		printk("temp: %d.%06d; press: %d.%06d; humidity: %d.%06d\n",
-		      temp.val1, temp.val2, press.val1, press.val2,
-		      humidity.val1, humidity.val2);
+		printk("temp: %s%d.%d; press: %s%d.%d; humidity: %s%d.%d\n",
+			PRIq_arg(temp_data->readings[0].temperature, 6, temp_data->shift),
+			PRIq_arg(press_data->readings[0].pressure, 6, press_data->shift),
+			PRIq_arg(hum_data->readings[0].humidity, 6, hum_data->shift));
```
Comment (laurenmurphyx64) [inline samples/sensor/bme280/src/main.c:110]: The temperature reading doesn't seem to have any problems, but I've noticed not so fun things about the pressure and humidity readings. When I flash the async sample, it gives readings in these ranges:

```
*** Booting Zephyr OS build v3.6.0-3333-gf1c882ead1c0 ***
Found device "bme280@77", getting sensor data
temp: 24.789062; press: 77.296875; humidity: 18.599609
temp: 24.789062; press: 77.125000; humidity: 18.626953
temp: 24.781250; press: 76.617187; humidity: 18.632812
temp: 24.773437; press: 76.281250; humidity: 18.632812
temp: 24.773437; press: 74.929687; humidity: 18.638671
temp: 24.773437; press: 75.93750; humidity: 18.638671
temp: 24.773437; press: 74.421875; humidity: 18.638671
temp: 24.773437; press: 73.742187; humidity: 18.638671
temp: 24.773437; press: 72.382812; humidity: 18.638671
temp: 24.773437; press: 72.554687; humidity: 18.632812
temp: 24.773437; press: 70.15625; humidity: 18.638671
temp: 24.781250; press: 70.15625; humidity: 18.632812
temp: 24.773437; press: 68.156250; humidity: 18.632812
temp: 24.781250; press: 67.984375; humidity: 18.632812
...
temp: 24.882812; press: 71.437500; humidity: 18.558593
temp: 24.882812; press: 70.757812; humidity: 18.552734
temp: 24.882812; press: 69.914062; humidity: 18.552734
temp: 24.882812; press: 69.742187; humidity: 18.558593
temp: 24.882812; press: 70.585937; humidity: 18.558593
temp: 24.882812; press: 71.601562; humidity: 18.558593
```
Then I flash the mainline non-async a second later, and get this:
```
��[00:00:00.000,000] <dbg> BME280: bme280_chip_init: ID OK
[00:00:00.010,000] <dbg> BME280: bme280_chip_init: "bme280@77" OK
[00:00:00.011,000] <err> FXOS8700: Could not get WHOAMI value
*** Booting Zephyr OS build v3.6.0-2349-gb9f3d68863b9 ***
Found device "bme280@77", getting sensor data
temp: 24.910000; press: 101.936925; humidity: 22.974609
temp: 24.900000; press: 101.937097; humidity: 22.997070
temp: 24.890000; press: 101.937265; humidity: 22.997070
temp: 24.890000; press: 101.938027; humidity: 22.974609
temp: 24.880000; press: 101.937687; humidity: 22.917968
temp: 24.880000; press: 101.938363; humidity: 22.906250
temp: 24.880000; press: 101.938617; humidity: 22.985351
temp: 24.880000; press: 101.938871; humidity: 23.029296
temp: 24.870000; press: 101.939378; humidity: 22.939453
```

File: samples/sensor/bme280/src/main.c (hunk 1 above)
Comment (teburd) [reply samples/sensor/bme280/src/main.c:27]: DT_COMPAT_GET_ANY_STATUS_OKAY(bosch_bme280) should return the same token needed here

File: samples/sensor/bme280/src/main.c (hunk 2 above)
Comment (teburd) [reply samples/sensor/bme280/src/main.c:110]: I don't believe the bme280 has an emulated device today, but that's exactly the sort of thing the emulated device would let you test for, that given the raw sensor value for channel X is set to some value, the resulting output from the sensor API works out to be the right thing. So you have some table of raw -> expected q31 values you can write a test that validates the conversions are correct. Otherwise this looks like an issue with the conversions somewhere still (these can be tricky).

File: samples/sensor/bme280/src/main.c (hunk 1 above)
Comment (laurenmurphyx64) [reply samples/sensor/bme280/src/main.c:27]: Changed

File: samples/sensor/bme280/src/main.c (hunk 2 above)
Comment (laurenmurphyx64) [reply samples/sensor/bme280/src/main.c:110]: So, I made some fixes to the humidity and pressure conversions that fixed pressure (?), but not humidity.

*mainline*
```
*** Booting Zephyr OS build v3.6.0-3733-g62ab24220572 ***
Found device "bme280@77", getting sensor data
temp: 25.040000; press: 100.918589; humidity: 29.324218
temp: 25.030000; press: 100.918867; humidity: 29.311523
temp: 25.020000; press: 100.919300; humidity: 29.266601
temp: 25.020000; press: 100.919312; humidity: 29.176757
temp: 25.010000; press: 100.919148; humidity: 29.188476
temp: 25.010000; press: 100.918898; humidity: 29.154296
temp: 25.010000; press: 100.919406; humidity: 29.154296
temp: 25.010000; press: 100.919746; humidity: 29.120117
temp: 25.020000; press: 100.920246; humidity: 29.121093
temp: 25.020000; press: 100.920156; humidity: 29.131835
```
*PR*
```
Found device "bme280@77", getting sensor data
temp: 25.109375; press: 100.929687; humidity: 4194185.193359
temp: 25.109375; press: 100.929687; humidity: 4194185.193359
temp: 25.109375; press: 100.929687; humidity: 4194185.195312
temp: 25.101562; press: 100.929687; humidity: 4194185.195312
temp: 25.101562; press: 100.929687; humidity: 4194185.195312
temp: 25.101562; press: 100.921875; humidity: 4194185.195312
temp: 25.101562; press: 100.921875; humidity: 4194185.195312
...
temp: 25.7812; press: 100.921875; humidity: 4194185.232421
temp: 24.992187; press: 100.921875; humidity: 4194185.238281
temp: 24.992187; press: 100.921875; humidity: 4194185.238281
temp: 24.976562; press: 100.921875; humidity: 4194185.240234
temp: 24.968750; press: 100.921875; humidity: 4194185.242187
temp: 24.960937; press: 100.921875; humidity: 4194185.246093
temp: 24.953125; press: 100.921875; humidity: 4194185.246093
temp: 24.953125; press: 100.921875; humidity: 4194185.248046
temp: 24.937500; press: 100.921875; humidity: 4194185.250000
temp: 24.929687; press: 100.921875; humidity: 4194185.250000
temp: 24.929687; press: 100.921875; humidity: 4194185.251953
temp: 24.921875; press: 100.921875; humidity: 4194185.255859
temp: 24.921875; press: 100.921875; humidity: 4194185.255859
```

What's weird is that the raw values coming from the sensor seem to be different for humidity on my PR compared to mainline, as in, the humidity raw values are way, way too big:

*mainline*
```
raw press 25833547 raw humidity 28759
temp: 25.060000; press: 100.912292; humidity: 28.084960
raw press 25833378 raw humidity 28655
temp: 25.050000; press: 100.911632; humidity: 27.983398
raw press 25833531 raw humidity 28667
temp: 25.050000; press: 100.912230; humidity: 27.995117
raw press 25833298 raw humidity 28677
temp: 25.040000; press: 100.911320; humidity: 28.004882
raw press 25833325 raw humidity 28699
temp: 25.030000; press: 100.911425; humidity: 28.026367
raw press 25833287 raw humidity 28698
temp: 25.020000; press: 100.911277; humidity: 28.025390
```

*PR*
```
raw press 25833836 raw humidity 4294845634
temp: 25.117187; press: 100.914062; humidity: 4194185.189453
raw press 25833665 raw humidity 4294845637
temp: 25.109375; press: 100.914062; humidity: 4194185.193359
raw press 25833385 raw humidity 4294845637
temp: 25.109375; press: 100.914062; humidity: 4194185.193359
raw press 25833277 raw humidity 4294845640
temp: 25.109375; press: 100.914062; humidity: 4194185.195312
raw press 25833343 raw humidity 4294845640
temp: 25.101562; press: 100.914062; humidity: 4194185.195312
raw press 25833387 raw humidity 4294845640
temp: 25.101562; press: 100.914062; humidity: 4194185.195312
```

When I intervene in bme_sample_fetch_helper and substitute a raw value whose decimal value I know in as the comp_humidity (47445 -> 46.333 RH, according to bme_channel_get() comments), it comes through the conversion exactly as expected:

```
> data->comp_humidity = 47445; // dev_data->comp_humidity;
< raw press 25832874 raw humidity 47445
   temp: 25.23437; press: 100.906250; humidity: 46.333984
```

**Comment by laurenmurphyx64** [comment]
Rebased with updates to https://github.com/zephyrproject-rtos/zephyr/pull/71160 and dropping https://github.com/zephyrproject-rtos/zephyr/pull/71093 commit as it has been merged. Ready for review.

**Comment by laurenmurphyx64** [comment]
Rebased now that https://github.com/zephyrproject-rtos/zephyr/pull/71160 is merged.

File: drivers/sensor/bosch/bme280/bme280_decoder.c
Comment (ubieda) [inline drivers/sensor/bosch/bme280/bme280_decoder.c:1]: Perhaps these changes belong to the previous commit? Or they deserve to be on a separate commit?

File: samples/sensor/bme280/src/main.c [hunk 3]
```diff
@@ -9,15 +9,27 @@
...
 #include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/sensor_data_types.h>
+#include <zephyr/rtio/rtio.h>
+#include <zephyr/dsp/print_format.h>
 
 /*
  * Get a device structure from a devicetree node with compatible
  * "bosch,bme280". (If there are multiple, just pick one.)
+ *
+ * TODO: Won't work as expected with multiple BME280 devices.
```
Comment (ubieda) [inline samples/sensor/bme280/src/main.c:20]: Is this TODO still valid?

File: drivers/sensor/bosch/bme280/bme280_async.c [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	if (rc != 0) {
+		LOG_ERR("Failed to get a read buffer of size %u bytes", min_buf_len);
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return rc;
+	}
+
+	edata = (struct bme280_encoded_data *)buf;
+	edata->header.timestamp = k_ticks_to_ns_floor64(k_uptime_ticks());
+
+	rc = bme280_sample_fetch_helper(dev, SENSOR_CHAN_ALL, &edata->data);
```
Comment (ubieda) [inline drivers/sensor/bosch/bme280/bme280_async.c:30]: Heads-up: This is a blocking call, similar to the ones referenced in #73676. Assuming the suggested fix goes in, this should be fine.

I'm assuming this will change to a non-blocking flow using RTIO bus drivers when we consider there's enough support to make the breaking change.

File: samples/sensor/bme280/src/main.c [hunk 5]
```diff
@@ -37,23 +49,65 @@ static const struct device *get_bme280_device(void)
...
+
+		decoder->decode(buf,
+				(struct sensor_chan_spec) {SENSOR_CHAN_PRESS, 0},
+				&press_fit, 1, press_buf);
+
+		struct sensor_q31_data *press_data =
+				(struct sensor_q31_data *) press_buf;
+
+		uint32_t hum_fit = 0;
+		uint8_t hum_buf[64];
```
Comment (ubieda) [inline samples/sensor/bme280/src/main.c:98]: Is there a scenario where this won't be `struct sensor_q31_data`? If not, I'd suggest directly using that data-type.

Same suggestion applies for `press_buf` and `temp_buf`.

File: drivers/sensor/bosch/bme280/bme280_decoder.c
Comment (laurenmurphyx64) [reply drivers/sensor/bosch/bme280/bme280_decoder.c:1]: They should be in the previous commit oops :/ Unless you feel strongly about putting them separately

Comment (ubieda) [reply drivers/sensor/bosch/bme280/bme280_decoder.c:1]: No strong opinion. Feel free to ignore this one :)

File: drivers/sensor/bosch/bme280/bme280_async.c (hunk 4 above)
Comment (teburd) [reply drivers/sensor/bosch/bme280/bme280_async.c:30]: yeah we *really* need to move forward with something here, I think after 3.7 some shim with k_work or p4wq would be useful to create a i2c/spi rtio shim so drivers can use the RTIO API against buses without having to worry about this. Its clear I think at this point we need it to move sensor drivers forward while we push on bus drivers to implement the API that would better suite the task.

File: samples/sensor/bme280/src/main.c (hunk 5 above)
Comment (teburd) [reply samples/sensor/bme280/src/main.c:96]: Good eye, the values to decode into should indeed by q31's

Comment (laurenmurphyx64) [reply samples/sensor/bme280/src/main.c]: No, and I've changed to use the data type directly.

**Review by ubieda — DISMISSED** [review]

File: drivers/sensor/bosch/bme280/bme280_async.c (hunk 4 above)
Comment (ubieda) [reply drivers/sensor/bosch/bme280/bme280_async.c:30]: I'll attempt to capture an RFC for this

File: drivers/sensor/bosch/bme280/bme280.h [hunk 6]
```diff
@@ -158,4 +160,60 @@ extern const struct bme280_bus_io bme280_bus_io_i2c;
...
+ * RTIO
+ */
+
+struct bme280_decoder_header {
+	uint64_t timestamp;
+} __attribute__((__packed__));
+
+struct bme280_encoded_data {
+	struct bme280_decoder_header header;
+	struct bme280_data data;
```
Comment (teburd) [inline drivers/sensor/bosch/bme280/bme280.h:205]: This could be the compensated comp_temp, comp_press, comp_humidity without the large number of compensation parameters today

File: drivers/sensor/bosch/bme280/bme280_async.c [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+	rc = rtio_sqe_rx_buf(iodev_sqe, min_buf_len, min_buf_len, &buf, &buf_len);
+	if (rc != 0) {
+		LOG_ERR("Failed to get a read buffer of size %u bytes", min_buf_len);
+		rtio_iodev_sqe_err(iodev_sqe, rc);
+		return rc;
+	}
+
+	edata = (struct bme280_encoded_data *)buf;
+	edata->header.timestamp = k_ticks_to_ns_floor64(k_uptime_ticks());
```
Comment (teburd) [inline drivers/sensor/bosch/bme280/bme280_async.c:28]: @yperess as we have talked about, k_uptime_ticks() does not have nearly the precision we are looking for, were you planning to still work on a sensor_timestamp() API for this sort of thing still?

File: drivers/sensor/bosch/bme280/bme280_decoder.c [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	struct sensor_q31_data *out = data_out;
+
+	out->header.base_timestamp_ns = edata->header.timestamp;
+	out->header.reading_count = 1;
+
+	switch (chan_spec.chan_type) {
+	case SENSOR_CHAN_AMBIENT_TEMP:
+		bme280_convert_signed_temp_raw_to_q31(edata->data.comp_temp,
```
Comment (teburd) [inline drivers/sensor/bosch/bme280/bme280_decoder.c:94]: Since we are only using the compensated values in the decoder, they should be the only values being passed along. The various compensation digX values and such can be kept private to the driver saving valuable memory, these values don't change each reading

File: drivers/sensor/bosch/bme280/bme280_async.c (hunk 4 above)
Comment (ubieda) [reply drivers/sensor/bosch/bme280/bme280_async.c:30]: Captured in #73869

File: drivers/sensor/bosch/bme280/bme280_async.c [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+
+#include "bme280.h"
+
+LOG_MODULE_DECLARE(BME280, CONFIG_SENSOR_LOG_LEVEL);
+
+int bme280_submit(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
```
Comment (teburd) [inline drivers/sensor/bosch/bme280/bme280_async.c:12]: This should check and validate that the channels to read are correct

See how the channel array can be obtained here https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/tdk/icm42688/icm42688_rtio.c#L49

Comment (laurenmurphyx64) [reply drivers/sensor/bosch/bme280/bme280_async.c:12]: Updated with a check

File: drivers/sensor/bosch/bme280/bme280_decoder.c (hunk 8 above)
Comment (laurenmurphyx64) [reply drivers/sensor/bosch/bme280/bme280_decoder.c]: Switched to this approach

**Review by teburd — DISMISSED** [review]

**Comment by teburd** [comment]
Looks good to me, needs to pass CI

**Review by ubieda — DISMISSED** [review]

**Review by ubieda — APPROVED** [review]

**Comment by laurenmurphyx64** [comment]
> Looks good to me, needs to pass CI

Fixed compliance, CI should pass

**Review by teburd — APPROVED** [review]

File: drivers/sensor/bosch/bme280/bme280.c
Comment (MaureenHelm) [inline drivers/sensor/bosch/bme280/bme280.c:1]: Surprised to see changes in the driver for a commit that claims to update the sample. Is this a rebasing mistake?

**Review by MaureenHelm — APPROVED** [review]

**Review by teburd — APPROVED** [review]
